### PR TITLE
[JIT][script] Switch to emitting ScriptModule for scripted and traced functions

### LINKED ISSUE #8867: [JIT] Normalize representation of traced and scripted functions/modules
Currently we have a mix of actual types when we script/trace a function/module:

```
import torch

@torch.jit.script
def script_fn(x):
    return x

print('*****Script function*****')
print(type(script_fn))
print(isinstance(script_fn, torch._C.GraphExecutor))

@torch.jit.trace(torch.zeros(3, 3))
def trace_fn(x):
    return x

print('*****Traced function*****')
print(type(trace_fn))
print(isinstance(trace_fn, torch._C.GraphExecutor))

class ScriptModule(torch.jit.ScriptModule):
    @torch.jit.script_method
    def forward(self, x):
        return x
sm = ScriptModule()

print('*****Script module*****')
print(type(sm))
print(isinstance(sm, torch._C.GraphExecutor))

class TracedModule(torch.nn.Module):
    def forward(self, x):
        return x
tm = torch.jit.trace(torch.zeros(3, 3))(TracedModule())

print('*****Traced module*****')
print(type(tm))
print(isinstance(tm, torch._C.GraphExecutor))

```

```
*****Script function*****
<class 'torch._C.GraphExecutor'>
True
*****Traced function*****
<class 'torch._C.GraphExecutor'>
True
*****Script module*****
<class '__main__.ScriptModule'>
False
*****Traced module*****
<class 'torch.jit.TopLevelTracedModule'>
False
```

This causes issues, for example, since we can only inline a `PythonValue` into a ScriptModule if it's an instance of GraphExecutor(https://github.com/pytorch/pytorch/blob/2b926aafb09575dd83ebf4fe2a76cbe239596f0b/torch/csrc/jit/script/init.cpp#L82). This causes problems in the case of calling Script/Traced modules from script functions -- they show up as `PythonOp`s.

Let's make it so that all of {script,traced} {function,module} produce a Module, and re-work the inlining logic to merge modules in all cases. This will greatly simplify the logic and make things easier to understand and maintain

### PR DESCRIPTION
**Pull request by jamesr66a**
Solves https://github.com/pytorch/pytorch/issues/8716 and closes https://github.com/pytorch/pytorch/issues/8867

This makes it so that all of {script, traced} {module, function} create ScriptModules and implements proper inlining between them. This also greatly simplifies things and makes clear that tracing is a way to convert regular Python into a ScriptModule

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -689,6 +697,9 @@ class TopLevelTracedModule(TracedModule):
     def forward(self, *args, **kwargs):
         return self._get_method('forward')(*args, **kwargs)
 
+    def _slow_forward(self, *args, **kwargs):
```
Comment (zdevito) [inline torch/jit/__init__.py]: What happened here?

File: torch/jit/__init__.py [hunk 2]
```diff
@@ -301,11 +301,15 @@ def wrapper(func):
...
 
         if isinstance(func, torch.nn.Module):
-            module = TopLevelTracedModule(func, **executor_options)
-            module._create_method_from_trace('forward', func, args)
-            return module
+            orig = func
         else:
-            return torch._C.GraphExecutor(func, args, **executor_options)
+            # traced functions become a method on an Empty module
+            orig = ScriptModule()
```
Comment (zdevito) [inline torch/jit/__init__.py]: Shouldn't this just be a blank module? Why does it have to be a script module?

File: torch/csrc/jit/script/init.cpp [hunk 3]
```diff
@@ -77,12 +77,17 @@ struct VISIBILITY_HIDDEN PythonValue : public SugaredValue {
...
 
-    // this python object might be a @trace or @script stand-alone function
+    // this python object might be a @trace or @script function/module
     // if so, inline the graph rather than calling the python
-    if(py::isinstance<GraphExecutor>(self)) {
-      GraphExecutor& ge = py::cast<GraphExecutor&>(self);
-      ensureSizeMatches(loc, ge.graph()->inputs().size(), inputs.size(), "arguments");
-      return packOutputs(*m.graph(),inlineCallTo(*m.graph(), *ge.graph(), inputs));
+
+    if(py::isinstance<Module>(self)) {
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: This code is unsafe: We keep ModuleValues separate from PythonValues because the only legit module calls are to submodules which are obtained via attribute lookup from self. Otherwise you can call a non-submodule whose parameters are not owned by the surrounding module. Those parameters can then go out of scope and crash the program. 

There needs to be a check that the "module" being accessed this way has no parameters before the call to emit_call_to is valid.

**Review by zdevito — COMMENTED** [review]
Looks good, a couple important comments.

File: torch/jit/__init__.py [hunk 4]
```diff
@@ -369,7 +373,11 @@ def _script_graph(fn, _frames_up=0):
 
 def script(fn, optimize=True, _frames_up=0):
     graph = _script_graph(fn, _frames_up=_frames_up + 1)
-    return functools.wraps(fn)(torch._C.GraphExecutor(graph, optimize))
+    mod = ScriptModule()
+    mod._create_method_from_graph('forward', graph)
+    # Forward docstrings
+    mod.__doc__ = fn.__doc__
+    return mod
```
Comment (apaszke) [inline torch/jit/__init__.py]: Is there any noticeable difference in cost of calling a `ScriptModule` vs a `GraphExecutor`?

Comment (jamesr66a) [reply torch/jit/__init__.py]: I just tried on my mac:

Benchmark code
```
import torch

@torch.jit.script
def a_fn(x, y):
    return x + y

import time

x = torch.rand(3, 3)
y = torch.rand(3, 3)
s = time.time()
for i in range(10000):
    a_fn(x, y)
print('time elapsed:', time.time() - s)
```

Calling a GraphExecutor
```
time elapsed: 0.12578201293945312
```

Calling a ScriptModule
```
time elapsed: 0.26899099349975586
```

Comment (jamesr66a) [reply torch/jit/__init__.py]: @apaszke I just pushed an update that reduces the overhead by doing the `forward()` method lookup in C++, thus we don't incur pybind11 overhead twice. New timing (same benchmark):

Before
```
time elapsed: 0.15010666847229004
```

After
```
time elapsed: 0.16915202140808105
```

File: torch/csrc/jit/script/init.cpp [hunk 5]
```diff
@@ -452,6 +460,23 @@ void initJitScriptBindings(PyObject* module) {
...
+        throw std::runtime_error("Attempted to call graph_for on a Module without a compiled forward()");
+      })
+      .def("_forward", [](Module& self, py::args args) {
+        // We implement this in C++ to avoid incurring the pybind11 dispatch
+        // overhead twice: once to call into the method lookup for "forward"
+        // and once to actually invoke the method.
+        //
+        // There is a thin wrapper on top of this method in the C++ version of
+        // ScriptModule.
+        auto inputs = createVariableTensorList(args);
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: It would be nice to have a single function that does these three lines for both forward and __call__ on Methods. We are going to update this site in the future to handle more data types, and having two site will add potential for divergence.

**Review by zdevito — APPROVED** [review]
Looks good. only minor nits.
